File: lib/persistent-storage-stack.ts
import * as backup from "aws-cdk-lib/aws-backup";
import * as cdk from "aws-cdk-lib";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as ssm from "aws-cdk-lib/aws-ssm";

import { Construct } from "constructs";
import { Environment } from "./environment";

export class PersistentStorageStack extends cdk.Stack {
  readonly backupVault: backup.BackupVault;
  readonly spreadsheetBucket: s3.Bucket;
  readonly emailSuppressionList: dynamodb.Table;
  readonly snapshotIdParameter: ssm.StringParameter;

  constructor(
    scope: Construct,
    id: string,
    props: cdk.StackProps & { environment: Environment }
  ) {
    super(scope, id, props);

    const environment = props.environment;

    let allowedCorsOrigins = [`https://${environment.subdomainName}`];
    if (!environment.production)
      allowedCorsOrigins.push("http://localhost:5000");

    this.spreadsheetBucket = new s3.Bucket(this, "SpreadsheetBucket", {
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      encryption: s3.BucketEncryption.S3_MANAGED,
      lifecycleRules: [
        {
          transitions: [
            {
              transitionAfter: cdk.Duration.days(90),
              storageClass: s3.StorageClass.INFREQUENT_ACCESS,
            },
          ],
        },
      ],
      cors: [
        {
          allowedMethods: [s3.HttpMethods.GET],
          allowedOrigins: allowedCorsOrigins,
        },
      ],
      autoDeleteObjects: environment.production ? false : true,
      removalPolicy: environment.production
        ? cdk.RemovalPolicy.RETAIN
        : cdk.RemovalPolicy.DESTROY,
    });

    this.emailSuppressionList = new dynamodb.Table(
      this,
      "EmailSuppressionList",
      {
        partitionKey: {
          name: "email",
          type: dynamodb.AttributeType.STRING,
        },
        billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
        removalPolicy: environment.production
          ? cdk.RemovalPolicy.RETAIN
          : cdk.RemovalPolicy.DESTROY,
      }
    );

    this.snapshotIdParameter = new ssm.StringParameter(
      this,
      "ServerStorageSnapshotIdParameter",
      { stringValue: environment.server.storage.snapshotId ?? "N/A" }
    );

    this.backupVault = new backup.BackupVault(this, "BackupVault", {
      removalPolicy: environment.production
        ? cdk.RemovalPolicy.RETAIN
        : cdk.RemovalPolicy.DESTROY,
    });
  }
}

File: nitecap/persistent_storage_stack.py
import aws_cdk as cdk
import aws_cdk.aws_backup as backup
import aws_cdk.aws_dynamodb as dynamodb
import aws_cdk.aws_s3 as s3
import aws_cdk.aws_ssm as ssm

from constructs import Construct
from .configuration import Configuration


class PersistentStorageStack(cdk.Stack):
    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        *,
        configuration: Configuration,
        **kwargs,
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)

        if configuration.production:
            removal_policy = cdk.RemovalPolicy.RETAIN
        else:
            removal_policy = cdk.RemovalPolicy.DESTROY

        allowed_cors_origin = [f"https://{configuration.domain_name}"]

        if not configuration.production:
            allowed_cors_origin.append("http://localhost:5000")

        self.spreadsheet_bucket = s3.Bucket(
            self,
            "SpreadsheetBucket",
            block_public_access=s3.BlockPublicAccess.BLOCK_ALL,
            encryption=s3.BucketEncryption.S3_MANAGED,
            lifecycle_rules=[
                s3.LifecycleRule(
                    transitions=[
                        s3.Transition(
                            transition_after=cdk.Duration.days(90),
                            storage_class=s3.StorageClass.INFREQUENT_ACCESS,
                        )
                    ]
                )
            ],
            cors=[
                s3.CorsRule(
                    allowed_methods=[s3.HttpMethods.GET],
                    allowed_origins=allowed_cors_origin,
                )
            ],
            auto_delete_objects=False if configuration.production else True,
            removal_policy=removal_policy,
        )

        self.email_suppression_list = dynamodb.Table(
            self,
            "EmailSuppressionList",
            partition_key={
                "name": "email",
                "type": dynamodb.AttributeType.STRING,
            },
            billing_mode=dynamodb.BillingMode.PAY_PER_REQUEST,
            removal_policy=removal_policy,
        )

        self.snapshot_id_parameter = ssm.StringParameter(
            self,
            "ServerStorageSnapshotIdParameter",
            string_value=configuration.server.storage.snapshot_id or "N/A",
        )

        self.backup_vault = backup.BackupVault(
            self,
            "BackupVault",
            removal_policy=removal_policy,
        )
